node/narwhal/src/gateway.rs

```diff
@@ -793,3 +794,128 @@ impl<N: Network> Gateway<N> {
+        pub dev: Option<u8>,
+        #[strategy(0..MAX_WORKERS)]
+        pub workers_count: u8,
+        pub worker_storage: StorageInput,
+    }
+
+    impl GatewayInput {
+        pub fn to_gateway(&self) -> Gateway<N> {
+            let committee = self.committee_input.to_committee().unwrap();
+            let account = self.node_validator.get_account();
+            let dev = match self.dev {
+                None => None,
+                Some(dev) => Some(dev as u16),
+            };
+            Gateway::new(Arc::new(RwLock::new(committee)), account, dev).unwrap()
+        }
+
+        pub async fn generate_workers(
+            &self,
+            gateway: &Gateway<N>,
+        ) -> (IndexMap<u8, Worker<N>>, IndexMap<u8, WorkerSender<N>>) {
+            // Construct a map of the worker senders.
+            let mut tx_workers = IndexMap::new();
+            let mut workers = IndexMap::new();
+
+            // Initialize the workers.
+            for id in 0..self.workers_count {
+                // Construct the worker channels.
+                let (tx_worker, rx_worker) = init_worker_channels();
+                // Construct the worker instance.
+                let mut worker = Worker::new(id, gateway.clone(), self.worker_storage.to_storage()).unwrap();
+                // Run the worker instance.
+                worker.run(rx_worker).await.unwrap();
+
+                // Add the worker and the worker sender to maps
+                workers.insert(id, worker);
+                tx_workers.insert(id, tx_worker);
+            }
+            (workers, tx_workers)
+        }
+    }
+
+    #[proptest]
+    fn gateway_initialization(input: GatewayInput) {
+        let account = input.node_validator.get_account();
+        let address = account.address();
+
+        let gateway = match input.dev {
+            Some(dev) => {
+                let gateway = input.to_gateway();
+                let tcp_config = gateway.tcp().config();
+                assert_eq!(tcp_config.listener_ip, Some(IpAddr::V4(Ipv4Addr::LOCALHOST)));
+                assert_eq!(tcp_config.desired_listening_port, Some(MEMORY_POOL_PORT + (dev as u16)));
+                gateway
+            }
+            None => {
+                let gateway = input.to_gateway();
+                let tcp_config = gateway.tcp().config();
+                assert_eq!(tcp_config.listener_ip, Some(IpAddr::V4(Ipv4Addr::UNSPECIFIED)));
+                assert_eq!(tcp_config.desired_listening_port, Some(MEMORY_POOL_PORT));
+                gateway
+            }
+        };
+        let tcp_config = gateway.tcp().config();
+        assert_eq!(tcp_config.max_connections, MAX_COMMITTEE_SIZE);
+        assert_eq!(gateway.account().address(), address);
+    }
+
+    #[proptest(async = "tokio")]
+    async fn gateway_start(#[filter(|x| x.dev.is_some())] input: GatewayInput) {
+        let Some(dev) = input.dev else { unreachable!() };
+        println!("{input:?}");
```

Comment:
may want to remove the println!

Reply:
:grinning: woops, good catch, removed now